### PR TITLE
CDAP-14770 JobQueueDataset migration

### PR DESCRIPTION
**Pull request by poornachandra**
Build - https://builds.cask.co/browse/CDAP-DUT6823
ITN - https://builds.cask.co/browse/IT-ITM-96

### CODE REVIEW

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/queue/JobQueueDataset.java [hunk 1]
```diff
@@ -16,141 +16,121 @@
...
+ * Row Key is in the following format for a job: <p/>
+ *     &lt;partition_id>&lt;scheduleId>&lt;generationI>&lt;rowType
+ * <ul>
+ *   <li>The &lt;partition_id> is a hash based upon the scheduleId</li>
+ *   <li>The &lt;generationId> is used to distinguish jobs for the same schedule in the queue</li>
+ *   <li>The &lt;rowType> is used to record the value stored in that row - job data, marked for deletion time, etc.</li>
+ * </ul>
  */
-public class JobQueueDataset extends AbstractDataset implements JobQueue {
+public class JobQueueDataset implements JobQueue {
```
Comment (jayfjin) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/queue/JobQueueDataset.java]: JobQueueTable?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/queue/JobQueueDataset.java [hunk 2]
```diff
@@ -16,141 +16,121 @@
...
-  private static final byte[] COL = new byte[] {'C'};
-  private static final byte[] TO_DELETE_COL = new byte[] {'D'};
-  private static final byte[] IS_OBSOLETE_COL = new byte[] {'O'};
-  private static final byte[] JOB_ROW_PREFIX = new byte[] {'J'};
-  private static final byte[] ROW_KEY_SEPARATOR = new byte[] {':'};
-  private static final byte[] MESSAGE_ID_ROW_PREFIX = new byte[] {'M'};
-
-  private static final int DEFAULT_NUM_PARTITIONS = 16;
+  private final StructuredTable jobQueueTable;
+  private final AppMetadataStore appMetadataStore;
```
Comment (jayfjin) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/queue/JobQueueDataset.java]: why is this needed?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/queue/JobQueueDataset.java [hunk 3]
```diff
@@ -330,21 +404,15 @@ int getPartition(ScheduleId scheduleId) {
...
    * @param topic the topic to lookup the last message id
    * @return the id of the last fetched message for this subscriber on this topic,
    *         or {@code null} if no message id was stored before
    */
-  public String retrieveSubscriberState(String topic) {
-    Row row = table.get(getRowKey(topic));
-    byte[] messageIdBytes = row.get(COL);
-    return messageIdBytes == null ? null : Bytes.toString(messageIdBytes);
+  public String retrieveSubscriberState(String topic) throws IOException {
+    return appMetadataStore.retrieveSubscriberState(topic, TMS_SUBSCRIBER_ID);
```
Comment (jayfjin) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/queue/JobQueueDataset.java]: I think these methods are already in app meta store, what's the use of having them here?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/queue/JobQueueDataset.java [hunk 4]
```diff
@@ -353,25 +421,7 @@ public String retrieveSubscriberState(String topic) {
...
-   *
-   * @param datasetFramework framework to add types and datasets to
-   */
-  public static void setupDatasets(DatasetFramework datasetFramework) throws IOException,
-    DatasetManagementException {
-    datasetFramework.addInstance(JobQueueDataset.class.getName(), Schedulers.JOB_QUEUE_DATASET_ID,
-                                 DatasetProperties.builder().
-                                   add(Constants.Scheduler.JOB_QUEUE_NUM_PARTITIONS, DEFAULT_NUM_PARTITIONS)
-                                   .build());
+  public void persistSubscriberState(String topic, String messageId) throws IOException {
```
Comment (jayfjin) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/queue/JobQueueDataset.java]: ditto

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/Schedulers.java [hunk 5]
```diff
@@ -80,17 +74,10 @@ public static String triggerKeyForProgramStatus(ProgramId programId, ProgramStat
...
-    try {
-      return DatasetsUtil.getOrCreateDataset(
-        context, dsFramework, JOB_QUEUE_DATASET_ID, JobQueueDataset.class.getName(),
-        () -> DatasetProperties.of(
-          Collections.singletonMap(Constants.Scheduler.JOB_QUEUE_NUM_PARTITIONS,
-                                   cConf.get(Constants.Scheduler.JOB_QUEUE_NUM_PARTITIONS))));
-    } catch (DatasetManagementException | IOException e) {
-      throw Throwables.propagate(e);
-    }
+  public static JobQueueDataset getJobQueue(StructuredTableContext context, CConfiguration cConf) {
```
Comment (jayfjin) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/Schedulers.java]: I think this belongs in the JobQueueDataset class

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/schedule/queue/JobQueueDatasetTest.java [hunk 6]
```diff
@@ -92,10 +67,12 @@
 /**
  * Tests for {@link JobQueueDataset}.
  */
-public class JobQueueDatasetTest {
+public abstract class JobQueueDatasetTest {
```
Comment (jayfjin) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/schedule/queue/JobQueueDatasetTest.java]: s/JobQueueDataSetTest/JobQueueTableTest?

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/schedule/queue/JobQueueDatasetTest.java [hunk 7]
```diff
@@ -600,4 +509,10 @@ public int compare(Job o1, Job o2) {
       jobIter.close();
     }
   }
+
+  public static JobQueueDataset createJobQueue(StructuredTableContext context) {
```
Comment (jayfjin) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/schedule/queue/JobQueueDatasetTest.java]: Use function in JobQueueDataset once the function in Schedulers is moved there?

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/schedule/queue/JobQueueDatasetTest.java [hunk 8]
```diff
@@ -16,26 +16,8 @@
 
 package co.cask.cdap.internal.app.runtime.schedule.queue;
 
-import co.cask.cdap.api.dataset.DatasetManagementException;
```
Comment (jayfjin) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/schedule/queue/JobQueueDatasetTest.java:19]: copyright 2019 in this file and throughout PR

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/schedule/queue/SqlJobQueueDatasetTest.java [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+ */
+public class SqlJobQueueDatasetTest extends JobQueueDatasetTest {
+  @ClassRule
+  public static final TemporaryFolder TEMP_FOLDER = new TemporaryFolder();
+
+  private static EmbeddedPostgres pg;
+
+  @BeforeClass
+  public static void setup() throws Exception {
+    CConfiguration cConf = CConfiguration.create();
```
Comment (jayfjin) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/schedule/queue/SqlJobQueueDatasetTest.java]: where is this cConf used? StorageModule still returns null for dataSource inside the structured table if we are trying to inject I believe

File: cdap-master/src/main/java/co/cask/cdap/data/tools/JobQueueDebugger.java [hunk 10]
```diff
@@ -169,22 +150,21 @@ private void scanPartition(int partition, boolean trace) {
...
-      this.numPartitions = jobQueue.getNumPartitions();
+      this.transactionRunner = transactionRunner;
+      this.numPartitions = cConf.getInt(Constants.Scheduler.JOB_QUEUE_NUM_PARTITIONS);
     }
 
-    private void printTopicMessageIds() throws TransactionFailureException {
-      transactional.execute(context -> {
+    private void printTopicMessageIds() {
+      TransactionRunners.run(transactionRunner, context -> {
         System.out.println("Getting notification subscriber messageIds.");
```
Comment (jayfjin) [inline cdap-master/src/main/java/co/cask/cdap/data/tools/JobQueueDebugger.java:165]: Should we use a logger instead of System.out.println?

File: cdap-app-fabric/src/main/java/co/cask/cdap/scheduler/ScheduleNotificationSubscriberService.java [hunk 11]
```diff
@@ -136,31 +136,31 @@ protected void shutDown() throws Exception {
...
-      // Intentional NO-OP until CDAP-14848 is resolved
+    protected void storeMessageId(StructuredTableContext context, String messageId) throws IOException {
+      getJobQueue(context).persistSubscriberState(getTopicId().getTopic(), messageId);
     }
 
     @Nullable
     @Override
     protected String loadMessageId(DatasetContext datasetContext) {
-      return getJobQueue(datasetContext).retrieveSubscriberState(getTopicId().getTopic());
+      return null;
```
Comment (jayfjin) [inline cdap-app-fabric/src/main/java/co/cask/cdap/scheduler/ScheduleNotificationSubscriberService.java]: On second thought, are there any usages of the loadMessage/storeMessage using datasetContext left? If not can we remove in all classes?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/queue/JobQueueDataset.java (hunk 2 above)
Comment (poornachandra) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/queue/JobQueueDataset.java]: This is used to load and store the TMS subscriber state

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/queue/JobQueueDataset.java (hunk 3 above)
Comment (poornachandra) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/queue/JobQueueDataset.java]: This is just a wrapper method around app metadata store retrieve.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/queue/JobQueueDataset.java (hunk 4 above)
Comment (poornachandra) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/queue/JobQueueDataset.java]: This is just a wrapper method around app metadata store persist

File: cdap-master/src/main/java/co/cask/cdap/data/tools/JobQueueDebugger.java (hunk 10 above)
Comment (poornachandra) [reply cdap-master/src/main/java/co/cask/cdap/data/tools/JobQueueDebugger.java:165]: This System.out is the output of the tool printed on the terminal. It does not get written to a log file.

File: cdap-app-fabric/src/main/java/co/cask/cdap/scheduler/ScheduleNotificationSubscriberService.java (hunk 11 above)
Comment (poornachandra) [reply cdap-app-fabric/src/main/java/co/cask/cdap/scheduler/ScheduleNotificationSubscriberService.java]: Good point - there are no more usages. This can be removed now

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/schedule/queue/SqlJobQueueDatasetTest.java (hunk 9 above)
Comment (poornachandra) [reply cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/schedule/queue/SqlJobQueueDatasetTest.java]: The cConf can be removed. The StorageModule will be fixed once have code to use inject the JDBC driver

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/queue/JobQueueDataset.java (hunk 4 above)
Comment (jayfjin) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/queue/JobQueueDataset.java]: can we remove this and just call app meta store wherever this is used? seems like this doesn't belong here

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/queue/JobQueueDataset.java (hunk 1 above)
Comment (poornachandra) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/queue/JobQueueDataset.java]: Done

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/schedule/queue/JobQueueDatasetTest.java (hunk 6 above)
Comment (poornachandra) [reply cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/schedule/queue/JobQueueDatasetTest.java]: Done

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/queue/JobQueue.java [hunk 12]
```diff
@@ -41,14 +42,7 @@
...
    */
   @Nullable
-  Job getJob(JobKey jobKey);
-
-  /**
-   * Creates a new Job in the queue or updates an existing Job.
-   *
-   * @param job the new job
-   */
-  void put(Job job);
```
Comment (yaojiefeng) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/queue/JobQueue.java:51]: Is there a reason this method is removed from the interface? I see it is still present in the JobQueueTable

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/Schedulers.java (hunk 5 above)
Comment (poornachandra) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/Schedulers.java]: Moved

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/schedule/queue/JobQueueTableTest.java [hunk 13]
```diff
@@ -0,0 +1,518 @@
+/*
+ * Copyright © 2017 Cask Data, Inc.
```
Comment (yaojiefeng) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/schedule/queue/JobQueueTableTest.java]: Should be 2019

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/schedule/queue/JobQueueTableTest.java [hunk 14]
```diff
@@ -0,0 +1,518 @@
...
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Tests for {@link JobQueueTable}.
+ */
+public abstract class JobQueueTableTest {
+  protected abstract TransactionRunner newTransactionRunner();
```
Comment (yaojiefeng) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/schedule/queue/JobQueueTableTest.java]: Better to move this method after all the variables

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/sql/RetryingSqlTransactionRunner.java [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+
+import java.util.concurrent.TimeUnit;
+import javax.sql.DataSource;
+
+/**
+ * Retries SQL operations in case they fail due to conflict.
+ * This class is based on {@link Transactions#createTransactionalWithRetry}.
+ */
+public class RetryingSqlTransactionRunner implements TransactionRunner {
+  private static final String TRANSACTION_CONFLICT_SQL_STATE = "40001";
```
Comment (yaojiefeng) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/sql/RetryingSqlTransactionRunner.java:34]: Is there a document or file we can learn about these state code, better to put a comment about where to see them

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/sql/RetryingSqlTransactionRunner.java [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+  @Override
+  public void run(TxRunnable runnable) throws TransactionException {
+    int retries = 0;
+    while (true) {
+      try {
+        transactionRunner.run(runnable);
+        break;
+      } catch (SqlTransactionException e) {
+        // Retry only transaction failure exceptions
+        if (TRANSACTION_CONFLICT_SQL_STATE.equals(e.getSqlException().getSQLState())) {
```
Comment (yaojiefeng) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/sql/RetryingSqlTransactionRunner.java:53]: Should we throw the original exception if the exception is not a transaction failure?

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/schedule/queue/JobQueueDatasetTest.java (hunk 7 above)
Comment (poornachandra) [reply cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/schedule/queue/JobQueueDatasetTest.java]: Done

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/queue/JobQueue.java (hunk 12 above)
Comment (poornachandra) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/queue/JobQueue.java:51]: Generally, no service should be writing the Jobs directly to the table. JobQueue has other methods that create a job when required.

File: cdap-app-fabric/src/main/java/co/cask/cdap/scheduler/ScheduleNotificationSubscriberService.java (hunk 11 above)
Comment (poornachandra) [reply cdap-app-fabric/src/main/java/co/cask/cdap/scheduler/ScheduleNotificationSubscriberService.java]: I notice that there are other places too that can now be refactored to remove the old transaction executor. But it will lead to lots of unrelated changes to this PR. Since it is a simple re-factoring, I'll leave it to be done when CDAP-14848 is handled

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/queue/JobQueueDataset.java (hunk 4 above)
Comment (poornachandra) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/queue/JobQueueDataset.java]: Added to a TODO to do it during refactoring for CDAP-14876

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/schedule/queue/JobQueueDatasetTest.java (hunk 8 above)
Comment (poornachandra) [reply cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/schedule/queue/JobQueueDatasetTest.java:19]: Done

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/schedule/queue/JobQueueTableTest.java (hunk 13 above)
Comment (poornachandra) [reply cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/schedule/queue/JobQueueTableTest.java]: Done

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/schedule/queue/JobQueueTableTest.java (hunk 14 above)
Comment (poornachandra) [reply cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/schedule/queue/JobQueueTableTest.java]: Done

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/sql/RetryingSqlTransactionRunner.java (hunk 15 above)
Comment (poornachandra) [reply cdap-data-fabric/src/main/java/co/cask/cdap/data2/sql/RetryingSqlTransactionRunner.java:34]: I got it from this page - https://www.postgresql.org/docs/9.6/transaction-iso.html. Added a comment to that effect

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/sql/RetryingSqlTransactionRunner.java (hunk 16 above)
Comment (poornachandra) [reply cdap-data-fabric/src/main/java/co/cask/cdap/data2/sql/RetryingSqlTransactionRunner.java:53]: `TransactionRunners` interface takes care of unwrapping the `TransactionException`.

**Comment by poornachandra** [comment]
@jayfjin @yaojiefeng Addressed the comments, please take another look

File: cdap-app-fabric/src/main/java/co/cask/cdap/scheduler/ScheduleNotificationSubscriberService.java (hunk 11 above)
Comment (jayfjin) [reply cdap-app-fabric/src/main/java/co/cask/cdap/scheduler/ScheduleNotificationSubscriberService.java]: sg, thanks

**Comment by jayfjin** [comment]
LGTM

**Review by jayfjin — APPROVED** [review]

**Comment by poornachandra** [comment]
Build is green, merging...
